power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -583,6 +584,41 @@ void add_tap_regulator_output(State const& state,
     }
 }
 
+template <class Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component> &&
+             std::disjunction_v<std::is_base_of<Transformer, Component>,
+                                std::is_base_of<ThreeWindingTransformer, Component>>
+constexpr void get_transformer_tap_positions(main_core::MainModelState<ComponentContainer> const& state,
+                                             TransformerTapPositionResult& transformer_tap_positions) {
+    constexpr auto group_index = ComponentContainer::template get_type_idx<Component>();
+    for (auto const& transformer : state.components.template citer<Component>()) {
+        transformer_tap_positions.push_back(
+            std::pair<Idx2D, IntS>{Idx2D{group_index, transformer.id()}, static_cast<IntS>(transformer.tap_pos())});
+    }
+}
+
+#ifdef _MSC_VER
+// MSVC compiler
+#pragma warning(push)
+#pragma warning(disable : 4717)
+#elif defined(__GNUC__) || defined(__clang__)
+// GCC or Clang compilers
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Winfinite-recursion"
+#endif
+template <typename... ComponentTypes, main_core::main_model_state_c State>
+inline void get_transformer_tap_positions(State const& state, TransformerTapPositionResult& transformer_tap_positions) {
+    using ComponentContainerType = typename State::ComponentContainer;
+    (get_transformer_tap_positions<ComponentTypes, ComponentContainerType>(state, transformer_tap_positions), ...);
+}
+#ifdef _MSC_VER
+// MSVC compiler
+#pragma warning(pop)
+#elif defined(__GNUC__) || defined(__clang__)
+// GCC or Clang compilers
+#pragma GCC diagnostic pop
+#endif
```

Comment:
this feels like it's extremely dangerous. it is highly likely that the compiler better knows what happens under the hood than we are - especially if all three compilers mention the same thing.

Even more so: the severity level is `\W1` for MSVC (critical). If this really is a false positive, please reproduce a minimal case in https://godbolt.org/ and file a bug report

I believe that the issue comes from the fact that the compiler does not understand that it should use the function in line 591 instead of this line because the signatures are different. Making the signatures the same should fix this

Reply:
Maybe discuss this in the next session. This seems to be very far.

Reply:
yeah we've already discussed that that we'll go over this together in a cleanup round later today

Reply:
I also don't think this would be the right way to implement it :)

Reply:
@petersalemink95 It most definitely isn't. We'll see what additional logic is needed for the compiler to know for sure.

Reply:
Issue solved with a new way of doing variadic calls.